### PR TITLE
Fix tests for transmission target branch

### PR DESCRIPTION
**Pull request by MatkovIvan**
Tests for #637

### CODE REVIEW

**Comment by guperrot** [comment]
@TroubleMakerBen looks good to me but since its going to land on your branch I let you merge.

File: sdk/appcenter/src/androidTest/java/com/microsoft/appcenter/AppCenterAndroidTest.java [hunk 1]
```diff
@@ -22,11 +22,7 @@
 import java.util.concurrent.Semaphore;
 import java.util.concurrent.atomic.AtomicReference;
 
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertNotEquals;
-import static org.junit.Assert.assertNotNull;
-import static org.junit.Assert.assertNull;
-import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.*;
```
Comment (guperrot) [inline sdk/appcenter/src/androidTest/java/com/microsoft/appcenter/AppCenterAndroidTest.java]: No star import, please always to the following with android commit dialog: organize imports, reformat, code analysis etc... Organize imports will automatically replace start with actual imports.

Comment (MatkovIvan) [reply sdk/appcenter/src/androidTest/java/com/microsoft/appcenter/AppCenterAndroidTest.java]: This is a change came from the base branch

Comment (guperrot) [reply sdk/appcenter/src/androidTest/java/com/microsoft/appcenter/AppCenterAndroidTest.java]: I commented on that PR too on other places, but since it appears green (new changes), it also applies on yours.

Comment (guperrot) [reply sdk/appcenter/src/androidTest/java/com/microsoft/appcenter/AppCenterAndroidTest.java]: Looks like a github bug, the full diff does not show it.

File: sdk/appcenter/src/androidTest/java/com/microsoft/appcenter/AppCenterAndroidTest.java [hunk 2]
```diff
@@ -23,15 +23,25 @@
...
+    private static final String DUMMY_APP_SECRET = "123e4567-e89b-12d3-a456-426655440000";
+
+    private static final String DUMMY_TRANSMISSION_TARGET_TOKEN = "snfbse234jknf";
+
     private Application mApplication;
 
     @Before
     public void setUp() throws IllegalAccessException, ClassNotFoundException, InstantiationException {
         AppCenter.unsetInstance();
+        DummyService.sharedInstance = null;
```
Comment (guperrot) [inline sdk/appcenter/src/androidTest/java/com/microsoft/appcenter/AppCenterAndroidTest.java]: unsetInstance is the pattern we use.

File: sdk/appcenter/src/androidTest/java/com/microsoft/appcenter/AppCenterAndroidTest.java [hunk 3]
```diff
@@ -78,9 +88,84 @@ public void setDefaultLogLevelDebug() throws IllegalAccessException, ClassNotFou
...
+    public void targetTokenWithAppSecretTest() {
+        String secret = AppCenter.TRANSMISSION_TARGET_TOKEN_KEY + AppCenter.KEY_VALUE_DELIMITER + DUMMY_TRANSMISSION_TARGET_TOKEN + AppCenter.PAIR_DELIMITER + DUMMY_APP_SECRET;
+        AppCenter.start(mApplication, secret, DummyService.class);
+        assertTrue(AppCenter.isEnabled().get());
+        verify(DummyService.getInstance()).onStarted(any(Context.class), eq(DUMMY_APP_SECRET), eq(DUMMY_TRANSMISSION_TARGET_TOKEN), any(Channel.class));
+    }
+
+    @Test
+    public void targetTokenWithUnKeyedAppSecretTest() {
+        String secret = AppCenter.TRANSMISSION_TARGET_TOKEN_KEY + AppCenter.KEY_VALUE_DELIMITER + DUMMY_TRANSMISSION_TARGET_TOKEN +  AppCenter.PAIR_DELIMITER + AppCenter.APP_SECRET_KEY + AppCenter.KEY_VALUE_DELIMITER + DUMMY_APP_SECRET;
```
Comment (guperrot) [inline sdk/appcenter/src/androidTest/java/com/microsoft/appcenter/AppCenterAndroidTest.java]: Import statics?

File: sdk/appcenter/src/androidTest/java/com/microsoft/appcenter/AppCenterAndroidTest.java (hunk 2 above)
Comment (MatkovIvan) [reply sdk/appcenter/src/androidTest/java/com/microsoft/appcenter/AppCenterAndroidTest.java]: Not for dummy ones in tests.
https://github.com/Microsoft/AppCenter-SDK-Android/blob/b4d77a54661fb55a5dd5ec997ab50818a6f0480d/sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java#L130-L131

Do you want to change it everywhere?

Comment (guperrot) [reply sdk/appcenter/src/androidTest/java/com/microsoft/appcenter/AppCenterAndroidTest.java]: Ah dummy ones, ok. Somehow I thought it was on the new transmission classes. Dummy is fine.

File: sdk/appcenter/src/androidTest/java/com/microsoft/appcenter/AppCenterAndroidTest.java [hunk 4]
```diff
@@ -72,15 +86,90 @@ public void accept(UUID uuid) {
...
+    @Test
+    public void targetTokenWithUnKeyedAppSecretTest() {
+        String secret = TRANSMISSION_TARGET_TOKEN_KEY + KEY_VALUE_DELIMITER + DUMMY_TRANSMISSION_TARGET_TOKEN + PAIR_DELIMITER + APP_SECRET_KEY + KEY_VALUE_DELIMITER + DUMMY_APP_SECRET;
+        AppCenter.start(mApplication, secret, DummyService.class);
+        assertTrue(AppCenter.isEnabled().get());
+        verify(DummyService.getInstance()).onStarted(any(Context.class), eq(DUMMY_APP_SECRET), eq(DUMMY_TRANSMISSION_TARGET_TOKEN), any(Channel.class));
+    }
+
+    @Test
+    public void unknownKeyTest() {
```
Comment (clpolet) [inline sdk/appcenter/src/androidTest/java/com/microsoft/appcenter/AppCenterAndroidTest.java]: Good test to have thanks.

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java [hunk 5]
```diff
@@ -200,6 +198,27 @@ public Void answer(InvocationOnMock invocation) throws Throwable {
 
         /* Mock network state helper. */
         when(NetworkStateHelper.getSharedInstance(any(Context.class))).thenReturn(mNetworkStateHelper);
+
+        /* Mock string splitter. */
+        whenNew(TextUtils.SimpleStringSplitter.class).withAnyArguments().then(new Answer<TextUtils.SimpleStringSplitter>() {
```
Comment (clpolet) [inline sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java]: I pushed a new implementation using pure java instead of Android string splitter here: feature/transmission-token-parsing. That way there is no need for mocking. Can you please take a look and merge it to this branch ?

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java [hunk 6]
```diff
@@ -442,9 +442,7 @@ private synchronized boolean instanceConfigure(Application application, String a
 
             /* A value with no key is default to the app secret. */
             if (!keyValueSplitter.hasNext()) {
-                if (keyValue.indexOf(KEY_VALUE_DELIMITER) == -1) {
```
Comment (clpolet) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: The reason behind this test was in the case of "appsecret=xxxxxxx;target=", the "target" would become the app secret without the extra check. But I think you are right in a way, no need for multiplying checks like that for corner cases we should instead validate the whole string via regex before parsing it.  If that regex check is really quick to implement for you then go ahead otherwise I'll create a task for it, just let me know.

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java (hunk 5 above)
Comment (clpolet) [reply sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java]: I should have told you about this implementation before, it was lying down in a local branch for more than a week :sweat_smile:.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java (hunk 6 above)
Comment (MatkovIvan) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: It didn't check anything. This check was fully equal to check above.

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java (hunk 5 above)
Comment (MatkovIvan) [reply sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java]: These PRs are targeted to different branches. I think we need to merge some of them before merging the changes between them.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java (hunk 6 above)
Comment (MatkovIvan) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/AppCenter.java]: Yes. I mean `if (!keyValueSplitter.hasNext())`  does the same thing

File: sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java (hunk 5 above)
Comment (clpolet) [reply sdk/appcenter/src/test/java/com/microsoft/appcenter/AppCenterTest.java]: True :/.

**Comment by clpolet** [comment]
Close because of duplicate work, will keep `feature/transmission-token-parsing` branch but we'll still port the few unit tests from here.
